go: document existence of orchestrion (#24302)
DataDog/documentation: content/en/tracing/trace_collection/automatic_instrumentation/compile_time/go.md

Author: Add a use-cases section

---
title: Automatic Compile-Time Instrumentation for Go
code_lang: go
further_reading:
- link: https://datadoghq.dev/orchestrion
  tag: Documentation
  text: Reference documentation
- link: https://github.com/DataDog/orchestrion/discussions
  tag: GitHub
  text: Discuss on GitHub
- link: https://github.com/DataDog/orchestrion/releases
  tag: GitHub
  text: Releases
- link: https://github.com/DataDog/orchestrion
  tag: GitHub
  text: Source code
---

<div class="alert alert-info">

  This feature is under active development.

  Early adopters are encouraged to provide feedback via any of the following channels:
  - [GitHub issues][1]
  - [This form][2]

>   - [GitHub discussions][3]

Author: Only the Google form here

## Introduction

Orchestrion is Datadog's automatic compile-time instrumentation tool for Go applications. It automatically adds instrumentation (backed by the [Go Datadog Trace library][4]) to Go applications during compilation.

Running at compile-time allows Orchestrion to not only instrument your own code, but also all of your dependencies' as
well as the Go standard library, ensuring you get the best coverage possible.

## Compatibility

### Go language

Orchestrion supports the two latest releases of Go, matching Go's [official release policy][5]. Older releases may function, but are not supported by default. For special requests, [contact support][6].

Orchestrion only supports [go modules][7].

### Framework Support

While the goal is to eventually support all integrations available in the [Go Datadog Trace library][4], support is currently only built for a gowing subset of them. The complete list of supported frameworks can be found in the [documentation][8].

## Getting Started

### Installation

1. Install Orchestrion in your environment:
   ```sh
   go install github.com/datadog/orchestrion@latest
   ```
2. If necessary, also add `GOBIN` to your `PATH`:
   ```sh

>    export PATH="${PATH}:$(go env GOBIN)"

Reviewer: `go env GOBIN` is empty by default, so it might be simpler to document `GOBIN=$PWD go install github.com/datadog/orchestrion@latest` and then `./orchestrion` everywhere?

Author: That sounds gross to me... Instead, I'll want to research where go installs when GOBIN is blank and how to discover it...

Also `./orchestrion` is likely not gonna work on Windows, which is another issue.

   ```
3. Register Orchestrion in your project's `go.mod` to ensure reproductibe builds:
   ```sh
   orchestrion pin
   ```
4. Commit changes to your version control system:
   ```sh
   git commit -m "chore: enable orchestrion" go.mod go.sum orchestrion.tool.go
   ```

### Usage

Adjust your build commands or environment to be Orchestrion-enabled using either of these options:

- **Use `orchestrion go`:** Simply prepend all your usual `go` commands with `orchestrion`. For example:
  ```sh
  orchestrion go build .
  orchestrion go run .
  orchestrion go test ./...
  ```

> - **Add a `-toolexec` argument:** Manually add the required `-toolexec="orchestrion toolexec"` argument to your `go` commands:

Reviewer: I think toolexec requires the fullpath, no?

Author: No it can get from `$PATH`

- **Modify `$GOFLAGS`:** Modify the `$GOFLAGS` environment variable to inject orchestrion, and use `go` commands normally:
  ```sh
  # Make sure to include the quotes as shown below, as these are required for
  # the Go toolchain to parse GOFLAGS properly!
  export GOFLAGS="${GOFLAGS} '-toolexec=orchestrion toolexec'"
  go build .
  go run .
  go test ./...
  ```

## Manual instrumentation

While Orchestrion automatically instruments your application for supported [integrations][4], there may be cases where
you still want to manuall add instrumentation to specific sections of your code.

### Custom trace spans

Apply the `//dd:span` comment on any function that accepts a `context.Context` or `*http.Request` argument, and Orchestrion will automatically create a trace span for all invocations of this function. This also works for function expressions.

```go
//dd:span custom_tag:tag_value
func CriticalPathFunction(ctx context.Context) {
  // ... implementation details ...
}
```

### Using the Tracer library

You can use the [Tracer library][4] in your application even if Orchestrion is in use. This can be used to instrument features that Orchestrion does not currently support.

> If Orchestrion get support of this feature at a later point, the manual call may result in duplicated spans being present in the traces, which cane asily be resolved by removing the manual calls that are no longer needed. Refer to Orchestrion [release notes][9] to know about new features.

Reviewer: ```suggestion
If Orchestrion gets support for this feature at a later point, the manual call may result in duplicated spans being present in the traces, which can easily be resolved by removing the manual calls that are no longer needed. Refer to Orchestrion [release notes][9] to learn about new features.
```

> If Orchestrion get support of this feature at a later point, the manual call may result in duplicated spans being present in the traces, which cane asily be resolved by removing the manual calls that are no longer needed. Refer to Orchestrion [release notes][9] to know about new features.

Reviewer: what about rather documenting right now to put manual instrum in separate files tagged with `dd:ignore` so you get something future-proof and safer directly?

Author: You can't tag a whole file, only a given AST node... unless we allow it on the `package` declaration (which is not a thing today). Also, that works for your own code, but not your dependencies (if these are already instrumented for Datadog)
